### PR TITLE
DUPLO-10189: v3 api s3  

### PR DESCRIPTION
**Pull request by pravin-duplocloud**
## Overview

DUPLO-10189: v3 api s3  

## Summary of changes

This PR does the following:

-  v3 api s3 
- fallback support to older api

## Testing performed

- [ ] Using unit tests
- [X] Manually, on my local system
- [X] Manually, on a remote test system

## Describe any breaking changes

- None

___

## **Type**
enhancement, bug_fix

___

## **Description**
- Added support for specifying the region of an S3 bucket in the Terraform provider.
- Integrated V3 API for S3 bucket operations, including create, read, update, and delete, with fallback to older API if V3 is not available.
- Enhanced error messages for clarity and improved debugging.
- Updated SDK to support new V3 API endpoints for S3 bucket operations.
- Updated documentation and examples to include the new `region` attribute for the S3 bucket resource.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_s3_bucket.go</strong><dd><code>Enhance S3 Bucket Resource with Region Support and V3 API Integration</code></dd></summary>
<hr>
      
duplocloud/resource_duplo_s3_bucket.go

<li>Added support for specifying the region of an S3 bucket.<br> <li> Implemented V3 API support for S3 bucket operations with fallback to <br>older API.<br> <li> Enhanced error messages for better clarity and debugging.<br> <li> Refactored S3 bucket creation, update, and deletion logic for improved <br>maintainability.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/431/files#diff-33c4307f72e0007e5c807624c3d14a7498e643ebb8d5d384cf357517b6f105f0">+216/-52</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tenant_aws_cloud_resources.go</strong><dd><code>Extend SDK with V3 API S3 Bucket Operations and Region Support</code></dd></summary>
<hr>
      
duplosdk/tenant_aws_cloud_resources.go

<li>Introduced new SDK functions for V3 API S3 bucket operations.<br> <li> Added <code>Region</code> field to <code>DuploS3Bucket</code> and <code>DuploS3BucketSettingsRequest</code> <br>structs.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/431/files#diff-92155358c5460b82348654a065ca4a047d7e3926595f74647c8786b1f7f02387">+65/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update CHANGELOG with S3 Bucket Enhancements and Fixes</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
CHANGELOG.md

<li>Updated CHANGELOG to reflect new features and fixes in the S3 bucket <br>resource.<br> <li> Cleaned up previous entries for better readability.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/431/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+26/-35</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>s3_bucket.md</strong><dd><code>Document Region Attribute for S3 Bucket Resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/resources/s3_bucket.md

<li>Documented the new <code>region</code> attribute for the S3 bucket resource.<br> <li> Provided examples on how to specify the region for an S3 bucket.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/431/files#diff-7799400e392f1817cdf3fcb4ebf58b5cfd17457bd5f918319e6e3bdfadbcd12c">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.tf</strong><dd><code>Add Example for Specifying S3 Bucket Region</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
examples/resources/duplocloud_s3_bucket/resource.tf

<li>Added an example Terraform configuration to demonstrate the use of the <br><code>region</code> attribute for an S3 bucket.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/431/files#diff-1dd3647865c2b7d81a278a579b829cd239a365c8295ed02f227290e2091b5827">+11/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-10189 TF: V3 APIs for S3](https://app.clickup.com/t/8655600/DUPLO-10189)

**Comment by pravin-duplocloud** [comment]
/update_changelog --pr_update_changelog.push_changelog_changes=true

File: duplocloud/resource_duplo_s3_bucket.go [hunk 1]
```diff
@@ -141,17 +147,36 @@ func resourceS3BucketRead(ctx context.Context, d *schema.ResourceData, m interfa
...
-	if duplo == nil {
-		d.SetId("") // object missing
-		return nil
-	}
+
+	// Figure out the full resource name.
+	features, _ := c.AdminGetSystemFeatures()
+	fullName, err := c.GetDuploServicesNameWithAws(tenantID, name)
 	if err != nil {
+		return diag.Errorf("Unable get duplo services name with aws, tenant %s bucket '%s': %s", tenantID, name, err)
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:156]:
```suggestion
		return diag.Errorf("Unable to get duplo service name for tenant %s and bucket '%s': %s", tenantID, name, err)
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 2]
```diff
@@ -141,17 +147,36 @@ func resourceS3BucketRead(ctx context.Context, d *schema.ResourceData, m interfa
...
 
+	// **** fallback on older api ****
+	if err != nil && err.PossibleMissingAPI() {
+		duplo, err = c.TenantGetS3BucketSettings(tenantID, name)
+		if duplo == nil {
+			d.SetId("") // object missing
+			return nil
+		}
+		if err != nil {
+			return diag.Errorf("Unable to retrieve tenant %s bucket '%s': %s", tenantID, name, err)
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:176]:
```suggestion
			return diag.Errorf("Unable to retrieve bucket '%s' for tenant %s: %s", name, tenantID, err)
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 3]
```diff
@@ -176,10 +196,152 @@ func resourceS3BucketCreate(ctx context.Context, d *schema.ResourceData, m inter
...
 	}
 	tenantID := d.Get("tenant_id").(string)
 
+	// prefix + name based on settings
+	fullName, errname := c.GetDuploServicesNameWithAws(tenantID, name)
+	if features.IsTagsBasedResourceMgmtEnabled {
+		fullName = name
+	}
+	if errname != nil {
+		return diag.Errorf("Error get duplo services name with aws '%s' error: '%s'", name, errname.Error())
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:209]:
```suggestion
		return diag.Errorf("Error retrieving duplo service name '%s' error: '%s'", name, errname.Error())
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 4]
```diff
@@ -176,10 +196,152 @@ func resourceS3BucketCreate(ctx context.Context, d *schema.ResourceData, m inter
...
+	}
+	errFill := fillS3BucketRequest(&duploObject, d)
+	if errFill != nil {
+		return diag.FromErr(errFill)
+	}
+
+	// Post the object to Duplo
+	_, err := c.TenantCreateV3S3Bucket(tenantID, duploObject)
+	if err != nil && !err.PossibleMissingAPI() {
+		return diag.Errorf("Error applying tenant %s bucket '%s': %s", tenantID, duploObject.Name, err)
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:224]:
```suggestion
		return diag.Errorf("Error creating bucket %s for tenant '%s': %s", duploObject.Name, tenantID, err)
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 5]
```diff
@@ -176,10 +196,152 @@ func resourceS3BucketCreate(ctx context.Context, d *schema.ResourceData, m inter
 		s3MaxLength = 63
 	}
 	if len(name) > s3MaxLength {
-		return diag.Errorf("Expected length of lambda function name '%s' to be in the range (1 - %d)", name, s3MaxLength)
+		return diag.Errorf("Expected length of s3 name '%s' to be in the range (1 - %d)", name, s3MaxLength)
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:199]:
```suggestion
		return diag.Errorf("Length of S3 bucket name '%s' must be in the range (1 - %d)", name, s3MaxLength)
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 6]
```diff
@@ -176,10 +196,152 @@ func resourceS3BucketCreate(ctx context.Context, d *schema.ResourceData, m inter
...
+	}
+	d.SetId(id)
+
+	duplo, err := c.TenantGetV3S3Bucket(tenantID, fullName)
+	if duplo == nil {
+		d.SetId("") // object missing
+		return nil
+	}
+	if err != nil {
+		return diag.Errorf("Unable to retrieve tenant %s bucket '%s': %s", tenantID, name, err)
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:248]:
```suggestion
		return diag.Errorf("Unable to retrieve bucket '%s' for tenant %s: %s", tenantID, name, err)
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 7]
```diff
@@ -176,10 +196,152 @@ func resourceS3BucketCreate(ctx context.Context, d *schema.ResourceData, m inter
...
+		return diag.FromErr(errName)
+	}
+
+	c := m.(*duplosdk.Client)
+	tenantID := d.Get("tenant_id").(string)
+
+	// Post the object to Duplo
+	resource, err := c.TenantUpdateV3S3Bucket(tenantID, duploObject)
+	if err != nil && !err.PossibleMissingAPI() {
+		return diag.Errorf("Error applying tenant %s bucket '%s': %s", tenantID, duploObject.Name, err)
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:280]:
```suggestion
		return diag.Errorf("Error updating bucket '%s' for tenant %s: %s", tenantID, duploObject.Name, err)
```

File: duplocloud/resource_duplo_s3_bucket.go [hunk 8]
```diff
@@ -176,10 +196,152 @@ func resourceS3BucketCreate(ctx context.Context, d *schema.ResourceData, m inter
...
+	log.Printf("[TRACE] resourceS3BucketCreate ******** start")
+	name := d.Get("name").(string)
+	// Create the request object.
+	duploObject := duplosdk.DuploS3BucketRequest{
+		Name:           name,
+		InTenantRegion: true,
+	}
+	//
+	c := m.(*duplosdk.Client)
+	//features, _ := c.AdminGetSystemFeatures()
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_s3_bucket.go:335]: delete unused code

File: duplosdk/tenant_aws_cloud_resources.go [hunk 9]
```diff
@@ -509,6 +511,68 @@ func (c *Client) TenantGetApplicationLB(tenantID string, name string) (*DuploApp
...
 
+// TenantCreateV3S3Bucket creates an S3 bucket resource via V3 Duplo Api.
+func (c *Client) TenantCreateV3S3Bucket(tenantID string, duplo DuploS3BucketSettingsRequest) (*DuploS3Bucket, ClientError) {
+
+	resp := DuploS3Bucket{}
+
+	// Create the bucket via Duplo.
+	err := c.postAPI(
+		fmt.Sprintf("TenantCreateV3S3Bucket(%s, %s)", tenantID, duplo.Name),
+		//  fmt.Sprintf("subscriptions/%s/S3BucketUpdate", tenantID),
```
Comment (ddiaz-duplo) [inline duplosdk/tenant_aws_cloud_resources.go:522]: remove commented out code

**Review by ddiaz-duplo — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_s3_bucket.go (hunk 1 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_s3_bucket.go:156]: done

File: duplocloud/resource_duplo_s3_bucket.go (hunk 2 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_s3_bucket.go:176]: done

File: duplocloud/resource_duplo_s3_bucket.go (hunk 3 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_s3_bucket.go:209]: done

File: duplocloud/resource_duplo_s3_bucket.go (hunk 4 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_s3_bucket.go]: done

File: duplocloud/resource_duplo_s3_bucket.go (hunk 6 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_s3_bucket.go]: done

File: duplocloud/resource_duplo_s3_bucket.go (hunk 8 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_s3_bucket.go]: done

File: duplosdk/tenant_aws_cloud_resources.go (hunk 9 above)
Comment (pravin-duplocloud) [reply duplosdk/tenant_aws_cloud_resources.go:522]: done

File: docs/resources/s3_bucket.md [hunk 10]
```diff
@@ -44,6 +44,8 @@ resource "duplocloud_s3_bucket" "www" {
   default_encryption {
     method = "Sse"
   }
+  # optional , if not provided, tenant region will be used
```
Comment (tahir-duplo) [inline docs/resources/s3_bucket.md:47]: Add this in example and generate doc using `make doc`

Comment (pravin-duplocloud) [reply docs/resources/s3_bucket.md]: done

**Review by tahir-duplo — APPROVED** [review]
LGTM

**Pull request merged**